lib/govuk_app_config/govuk_error/configuration.rb

```diff
@@ -112,6 +116,19 @@ def ignore_excluded_exceptions_in_data_sync
       }
     end
 
+    def strip_sensitive_data
+      lambda do |event, hint|
+        if hint[:exception] && contains_amqp_userinfo?(hint[:exception].message)
+          # rubocop:disable Style/HashEachMethods
+          event.exception.values.each do |exception|
+            exception.send(:instance_variable_set, "@value", remove_amqp_userinfo(exception.value))
+          end
+          # rubocop:enable Style/HashEachMethods
+        end
```

Comment:
Hello! Thank you very much for having a look. Have amended based on your other comments, but agree that this is the main thing. Have chucked in the example into the PR and commit message (https://sentry.io/organizations/govuk/issues/3798011523/?project=202242)

If there's a way to do this better, I'm super open to any ideas or different approaches entirely, as agreed that this doesn't feel super nice. Maybe we can have a chat.

Reply:
Thank you for sharing. Right so it's for `URI::InvalidURIError`, hmm that does seem delicate thing for us to try protect against as our ability to protect against that being written to Sentry is only based on our ability to predict the ways someone would get the URI wrong. So it might be hard to feel like this provides sufficient protection.

For example these are some URIs that this wouldn't catch:

"amqps://username:secret password@publishingmq.integration.govuk-internal.digital:5671/"
"amqps ://username:secretpassword@publishingmq.integration.govuk-internal.digital:5671/"
"amqps://username:secretpassword:publishingmq.integration.govuk-internal.digital:5671/"

Perhaps it might be a case that we want to catch URI::InvalidURIError and scrub that if it matches a very basic string - do we know if same situation can occur with other libraries that have URI based connection strings (such as Redis, PostgreSQL and MongoDB)? As it feels like it'd be inconsistent to only care about this for RabbitMQ if this was one of a few ways connection URI errors could make it to Sentry.

I guess the issue also gets logged in Logit?

There's a part of me, perhaps a purist side, that wonders if it's worth doing stuff when we're fighting low level and more to consider if you get the connection URI wrong and that gets spit out as an exception then you have to scrub the password, but I guess pragmatically someone wouldn't do that. I guess ideally libraries would also catch this error and wrap it to prevent it. So I guess if we are confident we _should_ deal with this error I feel it'd be better to target URI::InvalidURIError specifically

Reply:
I think you can do the filtering by setting event.message which may override the exception one (I've got to dash so haven't had a chance to test).

So I'm sort of expecting you could have something like:

```
next event unless hint[:exception].is_a?(URI::InvalidURIError)

match = hint[:exception].message.match(/(ampqs?:|postgresql:|mysql2:|redis:)/)

event.message = "Exception message filtered as it may be a #{match[0]} connection URI" if match

event
```

I could be over optimistic that this sufficies though.

Reply:
Also worth saying one option Al and I mentioned, but didn't discuss in detail, is not doing anything for this, so open to that being the conclusion.

Reply:
Yeah, I think in general it'd be very dubious of the library to log it in other scenarios, it's a bit annoying that the library doesn't catch this one as it could leak a credential. Unless we have evidence of other scenarios I wouldn't pursue them as it smells a bit like [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) though I can understand the motivation. You're probably right that there is another scenario it could happen in but I don't think we should write code unless we can evidence it as relatively likely.

So from my understanding of the Sentry code it only falls back to displaying the message from the exception if the message is blank: https://github.com/getsentry/sentry-ruby/blob/c7ba35c901dde97d28c9fc0a7a72e769525a5397/sentry-ruby/lib/sentry/event.rb#L75-L89. So I'm expecting that by message to something we override it. I'll try test it in the background of my next meeting.

Yup, sounds good about having an option of could before should - that makes a lot of sense. Our past precedence is to not have extra code for handling this - I'm imagining it's very rare that we've had invalid URIs in the past to demonstrate the need.

Reply:
So, as we both learnt my earlier optimism was misplaced and it wasn't as simple as changing the message :(

I came up with another approach, which borrows from their line in the Python docs of: "In Python a function can be used to modify the event or return a completely new one. If you return None, the event will be discarded."

I followed that and spun up a new Sentry::ErrorEvent instance for the issue so we can replace it with clean data. It's not near as clean as my earlier suggestion but is all public interfaces and shouldn't be leaky. It looks like:

```ruby
    def strip_sensitive_data
      lambda do |event, hint|
        next event unless hint[:exception].is_a?(URI::InvalidURIError)

        match = hint[:exception].message.match(/(ampqs?:|postgresql:|mysql2:|redis:)/)

        if match
          message = "Filtered URI::InvalidURIError due to potentially sensitive #{match[0]} connection URI"
          substitute_event = Sentry::ErrorEvent.new(
            configuration: event.configuration,
            message: message
          )
          substitute_event.add_threads_interface(backtrace: hint[:exception].backtrace)
          substitute_event
        else
          event
        end
      end
    end
```

I did find there's a bug in our code we need to change to for it to work though. In: https://github.com/alphagov/govuk_app_config/blob/13083f8f8563c4703c14fe8175ca35646ec64442/lib/govuk_app_config/govuk_error/configuration.rb#L125-L133 we need to change: `result = callback.call(event, hint)` to `result = callback.call(result, hint)` otherwise a subsequent before_send resets our new event.

Reply:
Oh and here's an example from Sentry: https://sentry.io/organizations/govuk/issues/3885353942/?project=202242&query=is%3Aunresolved&referrer=issue-stream (I've marked it as resolved). You can compare it to a non substituted_event in: https://sentry.io/organizations/govuk/issues/3885260848/?project=202242&query=is%3Aunresolved&referrer=issue-stream